evmcore/validation_test.go

```diff
@@ -0,0 +1,774 @@
+			}
+
+			setData(t, types.TxData(tx), maxInitCode)
+			setToNil(t, tx)
+
+			err := validateTx(types.NewTx(tx), types.NewPragueSigner(big.NewInt(1)),
+				testTransactionsOption())
+			require.ErrorIs(t, err, ErrMaxInitCodeSizeExceeded)
+		})
+	}
+
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("init code size not checked before shanghai/%v", name), func(t *testing.T) {
+			if isBlobOrSetCode(tx) {
+				t.Skip("blob and setCode transactions cannot be used before eip4844 and eip7702")
+			}
+			opt := testTransactionsOption()
+			opt.shanghai = false
+			opt.eip4844 = false
+			opt.eip7623 = false
+			// needs extra gas to allow big data to be afforded.
+			opt.currentMaxGas = 249_612
+
+			setData(t, types.TxData(tx), maxInitCode)
+			setToNil(t, tx)
+
+			setGasPriceOrFeeCap(t, tx, opt.currentBaseFee)
+			setGas(t, tx, opt.currentMaxGas) // enough gas
+			signer, address, signedTx := signTxForTest(t, tx)
+			testDb := newTestTxPoolStateDb()
+			testDb.balances[address] = uint256.NewInt(opt.currentMaxGas*opt.currentBaseFee.Uint64() + 1)
+			opt.currentState = testDb
+
+			err := validateTx(signedTx, signer,
+				opt)
+			require.NoError(t, err)
+		})
+	}
+
+}
+
+func TestValidateTx_Signer_RejectsTxWith(t *testing.T) {
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("unsigned tx/%v", name), func(t *testing.T) {
+			signer := types.HomesteadSigner{}
+			setSignatureValues(t, tx, big.NewInt(1), big.NewInt(2), big.NewInt(3))
+			err := validateTx(types.NewTx(tx), signer, testTransactionsOption())
+			require.ErrorIs(t, err, ErrInvalidSender)
+		})
+	}
+
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("invalid signer/%v", name), func(t *testing.T) {
+			signer := types.NewPragueSigner(big.NewInt(1))
+			setSignatureValues(t, tx, big.NewInt(1), big.NewInt(2), big.NewInt(3))
+			// sign txs with sender
+			key, err := crypto.GenerateKey()
+			require.NoError(t, err)
+			signedTx, err := types.SignTx(types.NewTx(tx), signer, key)
+			require.NoError(t, err)
+
+			// validate transaction
+			anotherSigner := types.NewPragueSigner(big.NewInt(2))
+			err = validateTx(signedTx, anotherSigner, testTransactionsOption())
+			require.ErrorIs(t, err, ErrInvalidSender)
+		})
+	}
+}
+
+func TestValidateTx_Balance_RejectsTxWhen(t *testing.T) {
+	for name, tx := range getTxsFromAllTypes() {
+		t.Run(fmt.Sprintf("insufficient balance/%v", name), func(t *testing.T) {
+			// setup validation context
+			opt := testTransactionsOption()
+
+			// --- needed for execution up to relevant check ---
+			// setup transaction enough gas and fee cap to reach balance check
+			setGasPriceOrFeeCap(t, tx, opt.currentBaseFee)
+			setGas(t, tx, opt.currentMaxGas)
+
+			// sign txs with sender
+			signer, address, signedTx := signTxForTest(t, tx)
+			// ---
+
+			// setup low balance
+			testDb := newTestTxPoolStateDb()
+			// balance = gas * fee cap + value
```

Comment:
Please also include a positive `value` in the transaction to make sure it is considered in the validation.

Reply:
https://github.com/0xsoniclabs/sonic/pull/154/commits/4205286062d9a53c041a02d2ccb9f0719ab7ba0f

Reply:
I do not think this was resolved. I do not see how the `value` would not be zero, and the linked commit touches changes not related to this.

Reply:
https://github.com/0xsoniclabs/sonic/commit/4205286062d9a53c041a02d2ccb9f0719ab7ba0f#diff-c2357ac27ddbb583bf7649ef1daa11ce9cd1b47a6c1c0e9483621f931c3d8138R449
this is the specific line from the linked committee that adds the value into the formula. 
But as you mentioned, all txs have a `value` of 0, I will change this, to a non zero value.

Reply:
https://github.com/0xsoniclabs/sonic/pull/154/commits/2bcc9f0c1d81257091eec396d0f790809ad9eae1
in this commit all transactions get assigned a non zero value.